src/rust/inetstack/protocols/layer4/tcp/established/sender.rs

```diff
@@ -393,68 +402,96 @@ impl Sender {
+            let mut bytes_remaining: usize = bytes_acknowledged as usize;
+            // Remove bytes from the unacked queue.
+            while bytes_remaining != 0 {
+                if let Some(mut segment) = self.unacked_queue.try_pop() {
+                    // Add sample for RTO if we have an initial transmit time.
+                    // Note that in the case of repacketization, an ack for the first byte is enough for the time sample.
+                    // TODO: TCP timestamp support.
+                    if let Some(initial_tx) = segment.initial_tx {
+                        self.rto_calculator.add_sample(now - initial_tx);
+                    }
+
+                    if segment.bytes.len() > bytes_remaining {
+                        // Only some of the data in this segment has been acked.  Remove just the acked amount.
+                        expect_ok!(
+                            segment.bytes.adjust(bytes_remaining),
+                            "'segment' should contain at least 'bytes_remaining'"
+                        );
+                        segment.initial_tx = None;
+
+                        // Leave this segment on the unacknowledged queue.
+                        self.unacked_queue.push_front(segment);
+                        break;
+                    }
+
+                    if segment.bytes.len() == 0 {
+                        // This buffer is the end-of-send marker.  So we should only have one byte of acknowledged
+                        // sequence space remaining (corresponding to our FIN).
+                        debug_assert_eq!(bytes_remaining, 1);
+                        bytes_remaining = 0;
+                    }
 
-                if segment.bytes.len() == 0 {
-                    // This buffer is the end-of-send marker.  So we should only have one byte of acknowledged sequence
-                    // space remaining (corresponding to our FIN).
-                    debug_assert_eq!(bytes_remaining, 1);
-                    bytes_remaining = 0;
+                    bytes_remaining -= segment.bytes.len();
+                } else {
+                    debug_assert!(false); // Shouldn't have bytes_remaining with no segments remaining in unacked_queue.
                 }
+            }
 
-                bytes_remaining -= segment.bytes.len();
-            } else {
-                debug_assert!(false); // Shouldn't have bytes_remaining with no segments remaining in unacked_queue.
+            // Update SND.UNA to SEG.ACK.
+            self.send_unacked.set(header.ack_num);
+
+            // Update send window.
+            // Check that the ACK we're using to update the window isn't older than the last one used to update it.
+            if self.send_window_last_update_seq < header.seq_num
+                || (self.send_window_last_update_seq == header.seq_num
+                    && self.send_window_last_update_ack <= header.ack_num)
```

Comment:
It would be good to extract this into a separate function or variable with a meaningful name.